whisper/whisperv5/filter.go

```diff
@@ -39,42 +41,67 @@ type Filter struct {
 }
 
 type Filters struct {
-	id       uint32 // can contain any value except zero
-	watchers map[uint32]*Filter
+	watchers map[string]*Filter
 	whisper  *Whisper
 	mutex    sync.RWMutex
 }
 
 func NewFilters(w *Whisper) *Filters {
 	return &Filters{
-		watchers: make(map[uint32]*Filter),
+		watchers: make(map[string]*Filter),
 		whisper:  w,
 	}
 }
 
-func (fs *Filters) Install(watcher *Filter) uint32 {
+func (fs *Filters) generateRandomID() (id string) {
+	var err error
+	buf := make([]byte, 20)
+	for i := 0; i < 3; i++ {
+		_, err = crand.Read(buf)
+		if err != nil {
+			continue
+		}
+		if !validateSymmetricKey(buf) {
+			err = fmt.Errorf("error in generateRandomID: crypto/rand failed to generate random data")
+			continue
+		}
+		id = common.Bytes2Hex(buf)
+		if fs.watchers[id] != nil {
+			err = fmt.Errorf("error in generateRandomID: generated same ID twice")
+			continue
+		}
+		return id
+	}
+
+	// at this point all the attempts to generate the random ID failed.
+	// this poses a serious danger, since whisper very much relies
+	// on random numbers. the program must be terminated.
+	panic(err)
```

Comment:
Do not use `panic`. Handle termination of **whisper** otherwise. Panic'ing forces the whole application to shutdown including other services.

As a side note, anything that `panic`s should be called `must` e.g. `mustGenerateRandomID()`.